### PR TITLE
Implement KVStore API

### PR DESCRIPTION
**Pull request by conopt**

### CODE REVIEW

File: include/optimizer.h [hunk 1]
```diff
@@ -53,15 +54,22 @@ class Optimizer {
...
+  void Update(int index, NDArray weight, NDArray grad);
   // TODO(zhangcheng-qinyinghua)
   // implement Update a list of arrays, maybe in the form of map
   //void Update(int index, std::vector<NDArray> weights, std::vector<NDArray> grad, mx_float lr);
 
+  /*!
+  *  \brief Serialize the optimizer parameters to a string.
+  *  \return serialization
+  */
+  std::string Serialize() const;
```
Comment (hjk41) [inline include/optimizer.h:69]: can we serialize into a byte array instead of a string? FP values can become different when you encode and then decode it from string.

File: include/optimizer.h [hunk 2]
```diff
@@ -53,15 +54,22 @@ class Optimizer {
...
+
 private:
   bool init_;
+  float learning_rate_;
+  std::string opt_type_;
```
Comment (hjk41) [inline include/optimizer.h:74]: I would rather use an enum type than using string. With enum type you can easily know what values you can use. With string, you will need to look up from the manual

**Comment by hjk41** [comment]
@zhangchen-qinyinghua This patch also modifies optimizer. Do you have any comments?

File: include/optimizer.h (hunk 1 above)
Comment (conopt) [inline include/optimizer.h:69]: The parameter assigned to optimizer will also be converted to string (see https://github.com/dmlc/MXNet.cpp/blob/202ec2daeb2fdd70c27b2134dc0dfe12cbf3d125/include/optimizer.h#L43), so byte encoding makes no difference here.

File: include/optimizer.h (hunk 2 above)
Comment (conopt) [inline include/optimizer.h:74]: Optimizer types are dynamically registered in runtime, so string is the only option.

File: include/optimizer.hpp [hunk 3]
```diff
@@ -29,10 +30,21 @@ void Optimizer::Update(int index, NDArray weight, NDArray grad, mx_float lr) {
       param_values.data(), &handle_);
     init_ = true;
   }
-  MXOptimizerUpdate(handle_, index, weight.GetHandle(), grad.GetHandle(), lr);
+  MXOptimizerUpdate(handle_, index, weight.GetHandle(), grad.GetHandle(), learning_rate_);
```
Comment (zhangchen-qinyinghua) [inline include/optimizer.hpp]: This usage comes from the old version c_api.h.
Please change it to the newer ones.

```
  MXNET_DLL int MXOptimizerUpdate(OptimizerHandle handle,
                                  int index,                                                                                                                 
                                  NDArrayHandle weight,
                                  NDArrayHandle grad,
                                  mx_float lr,
                                  mx_float wd); 
```

File: include/optimizer.hpp [hunk 4]
```diff
@@ -13,11 +13,12 @@
 namespace mxnet {
 namespace cpp {
 
-Optimizer::Optimizer(const std::string &opt_type) {
+Optimizer::Optimizer(const std::string &opt_type, mx_float learning_rate)
+  :init_(false), learning_rate_(learning_rate), opt_type_(opt_type) {
   MXOptimizerFindCreator(opt_type.c_str(), &creator_);
-  init_ = false;
```
Comment (zhangchen-qinyinghua) [inline include/optimizer.hpp:18]: I think this line is necessary, sometimes the bool is not initialized well.

File: include/optimizer.h [hunk 5]
```diff
@@ -24,8 +24,9 @@ class Optimizer {
   /*!
   * \brief Operator constructor, the optimizer is not initialized until the first Update
   * \param opt_type type of the optimizer
+  * \param learning_rate
   */
-  explicit Optimizer(const std::string &opt_type);
+  Optimizer(const std::string &opt_type, mx_float learning_rate);
```
Comment (zhangchen-qinyinghua) [inline include/optimizer.h]: Do we really need to give a learning_rate at the initialization?
If so, the weight decay is also needed.
And I think it better to make learning rate and weight decay mutable during the update.

File: include/optimizer.hpp (hunk 4 above)
Comment (hjk41) [inline include/optimizer.hpp:18]: init_ is already initiated in the initialize list

File: include/optimizer.h (hunk 5 above)
Comment (hjk41) [inline include/optimizer.h]: I think some optimizers might change the learning rate dynamically? Why do we want to specify the learning rate everytime we call update?

File: include/optimizer.h (hunk 2 above)
Comment (hjk41) [inline include/optimizer.h:74]: But we only have a few different optimizers, right? You can refer to how op.h handles enumerations. We can always map the enumeration value back to string before we pass them to the C api.

**Comment by hjk41** [comment]
@lx75249 please update the code to resolve conflicts

File: include/optimizer.h (hunk 5 above)
Comment (conopt) [inline include/optimizer.h]: In distributed setting, it's server's responsibility to update parameters with optimizer, and the [prototype](https://github.com/dmlc/MXNet.cpp/blob/master/include/c_api.h#L1071) of updater determines that the optimizer should maintain internally all the parameters including lr and wd. 
Current python [implementation](https://github.com/dmlc/mxnet/blob/master/python/mxnet/optimizer.py#L276) uses the same strategy. If the worker does need to control lr or wd, I can add setters for them in optimizer.

Comment (zhangchen-qinyinghua) [inline include/optimizer.h]: I think setters are good.
Sometimes we just want to modify the learning rate to some special layers by hand.

File: include/optimizer.hpp (hunk 4 above)
Comment (zhangchen-qinyinghua) [inline include/optimizer.hpp:18]: Oh my...
I shouldn't have ignored that..

**Comment by conopt** [comment]
Conflicts resolved.
